src/Enumerators/Enumerators.dfy

```diff
@@ -0,0 +1,520 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+include "../Frames.dfy"
+include "../Wrappers.dfy"
+include "../Collections/Sequences/Seq.dfy"
+
+module Enumerators {
+
+  import opened Frames
+  import opened Wrappers
+  import opened Seq
+
+
+  // A trait for any value that produces a finite set of value.
+  trait {:termination false} Enumerator<T> extends Validatable {
+    
+    // The Valid() predicate from the Validatable trait ends up
+    // becoming the "enumeration invariant", which in turn becomes
+    // the loop invariant in a while loop that uses an enumerator.
+
+    // All values produced by the Next() method in the order they
+    // were produced.
+    ghost var enumerated: seq<T>
+
+    // Any enumerator that produces one value at a time
+    // and provably terminates is equivalent to an enumerator
+    // that produces a specific seq<T>. This value may be underspecified
+    // such that it is not known, even its length, until after all
+    // values have been produced.
+    // Dafny doesn't let you pass around an underspecified value though,
+    // so we don't define a "to be enumerated" field or function.
+    
+    // The termination measure for the enumerator. Must decrease on every
+    // call to Next().
+    //
+    // Would be better as an arbitrary termination clause somehow instead,
+    // but we don't have language-level access to the built-in well-founded
+    // ordering. See https://github.com/dafny-lang/dafny/issues/762.
+    function Decreases(): nat
+      reads Repr
+      decreases Repr, 1
+      requires Valid()
+
+    // Pre-condition for Next(). Making this a pure predicate means that
+    // enumerators have to at least know ahead of time if they are able to
+    // produce a value, even if they do not know what value it is until
```

Comment:
I'm not convinced this is the best model for enumerators. Always having to compute a value ahead of requests makes the code more complicated. And it may also be confusing, because any side effects of computing a value take place when the previous value is in the process of being returned.

Reply:
Completely agree on those downsides. My strongest reason for landing on this version so far is that it was FAR easier to make all this code verify. Now that I've reached this baseline, though, I'll make an alternate version of all of this where `Next()` returns an `Option<T>` instead and see how it compares on both the implementation and consuming sides.